File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/SessionClock.swift
import Foundation

// MARK: - SessionClock
//
// Pause-aware elapsed time for a practice session, extracted from
// MindfulnessSessionManager so the bookkeeping can be tested without a watch,
// a Task, or a haptic engine.
//
// The manager kept this as two mutable fields — `accumulated` and
// `segmentStartedAt` — updated from five places (start, pause, resume, endEarly,
// the runner's completion). Every one of them had to get the arithmetic right,
// and nothing checked that they did. A pause that forgets to fold the running
// segment into the total silently shortens the session the user is credited
// with; one that folds it in twice silently lengthens it.

public struct SessionClock: Equatable, Sendable {

    /// Practice time banked by segments that have already ended.
    public private(set) var accumulated: TimeInterval
    /// When the current running segment began, or nil while paused.
    public private(set) var segmentStartedAt: Date?

    public init(accumulated: TimeInterval = 0, segmentStartedAt: Date? = nil) {
        self.accumulated = accumulated
        self.segmentStartedAt = segmentStartedAt
    }

    public var isRunning: Bool { segmentStartedAt != nil }

    /// Seconds of actual practice at `date`.
    ///
    /// Clamped at zero on the running segment: a backwards clock adjustment
    /// must not subtract from time the user really spent.
    public func elapsed(at date: Date) -> TimeInterval {
        guard let segmentStartedAt else { return accumulated }
        return accumulated + max(0, date.timeIntervalSince(segmentStartedAt))
    }

    public func remaining(at date: Date, of planned: TimeInterval) -> TimeInterval {
        max(0, planned - elapsed(at: date))
    }

    public func isComplete(at date: Date, of planned: TimeInterval) -> Bool {
        elapsed(at: date) >= planned
    }

    // MARK: Transitions
    //
    // Each returns a new clock rather than mutating in place, so a caller
    // cannot half-apply one.

    public static func started(at date: Date) -> SessionClock {
        SessionClock(accumulated: 0, segmentStartedAt: date)
    }

    /// Folds the running segment into the total and stops the clock.
    /// Pausing an already-paused clock changes nothing.
    public func paused(at date: Date) -> SessionClock {
        guard isRunning else { return self }
        return SessionClock(accumulated: elapsed(at: date), segmentStartedAt: nil)
    }

    /// Restarts the clock without touching what was already banked.
    /// Resuming a running clock changes nothing — in particular it does not
    /// discard the segment in flight.
    public func resumed(at date: Date) -> SessionClock {
        guard !isRunning else { return self }
        return SessionClock(accumulated: accumulated, segmentStartedAt: date)
    }

    /// Stops the clock, banking whatever was running.
    public func stopped(at date: Date) -> SessionClock {
        SessionClock(accumulated: elapsed(at: date), segmentStartedAt: nil)
    }

    /// Stops at exactly the planned duration — the natural-completion case,
    /// where the runner has already decided the session is over and the credited
    /// time should be the plan rather than however late the last tick landed.
    public func completed(of planned: TimeInterval) -> SessionClock {
        SessionClock(accumulated: planned, segmentStartedAt: nil)
    }
}

File: backend/infra/lambda/aria_core/session_clock.py
"""Pause-aware elapsed time for a practice session.

Python port of ForgeCore's ``SessionClock.swift``. Each transition returns a
new clock so a caller cannot half-apply one. WidgetKit / Watch haptics stay
on the client.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone


def _epoch(dt: datetime) -> float:
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.timestamp()


@dataclass(frozen=True)
class SessionClock:
    accumulated: float = 0.0
    segment_started_at: datetime | None = None

    @property
    def is_running(self) -> bool:
        return self.segment_started_at is not None

    def elapsed(self, at: datetime) -> float:
        if self.segment_started_at is None:
            return self.accumulated
        return self.accumulated + max(0.0, _epoch(at) - _epoch(self.segment_started_at))

    def remaining(self, at: datetime, planned: float) -> float:
        return max(0.0, planned - self.elapsed(at))

    def is_complete(self, at: datetime, planned: float) -> bool:
        return self.elapsed(at) >= planned

    @staticmethod
    def started(at: datetime) -> SessionClock:
        return SessionClock(accumulated=0.0, segment_started_at=at)

    def paused(self, at: datetime) -> SessionClock:
        if not self.is_running:
            return self
        return SessionClock(accumulated=self.elapsed(at), segment_started_at=None)

    def resumed(self, at: datetime) -> SessionClock:
        if self.is_running:
            return self
        return SessionClock(accumulated=self.accumulated, segment_started_at=at)

    def stopped(self, at: datetime) -> SessionClock:
        return SessionClock(accumulated=self.elapsed(at), segment_started_at=None)

    def completed(self, planned: float) -> SessionClock:
        return SessionClock(accumulated=planned, segment_started_at=None)
